File: src/com/suushiemaniac/cubing/bld/model/source/AlgSource.java
package com.suushiemaniac.cubing.bld.model.source;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType;

import java.net.URI;
import java.util.Set;

public interface AlgSource {
    boolean isReadable();

    boolean isWritable();

    URI getSourceURI();

    boolean mayRead();

    Set<Algorithm> getAlgorithms(PieceType type, String letterPair);

    Set<String> getRawAlgorithms(PieceType type, String letterPair);

    boolean mayWrite();

    boolean addAlgorithm(PieceType type, String letterPair, Algorithm algorithm);

    boolean addAlgorithms(PieceType type, String letterPair, Set<Algorithm> algorithms);

    boolean mayUpdate();

    boolean updateAlgorithm(PieceType type, Algorithm oldAlg, Algorithm newAlg);

    boolean mayDelete();

    boolean deleteAlgorithm(PieceType type, Algorithm algorithm);

    boolean deleteAlgorithms(PieceType type, String letterPair);
}

File: src/com/suushiemaniac/cubing/bld/model/source/AlgSource.kt
package com.suushiemaniac.cubing.bld.model.source

import com.suushiemaniac.cubing.alglib.alg.Algorithm
import com.suushiemaniac.cubing.bld.model.cycle.PieceCycle
import com.suushiemaniac.cubing.bld.model.enumeration.piece.PieceType

import java.net.URI

interface AlgSource {
    val isReadable: Boolean

    val isWritable: Boolean

    val sourceURI: URI

    fun mayRead(): Boolean

    fun getAlgorithms(type: PieceType, case: PieceCycle): Set<Algorithm> {
        return this.getRawAlgorithms(type, case).map(type.reader::parse).toSet()
    }

    fun getRawAlgorithms(type: PieceType, case: PieceCycle): Set<String>

    fun mayWrite(): Boolean

    fun addAlgorithm(type: PieceType, case: PieceCycle, algorithm: Algorithm): Boolean

    fun addAlgorithms(type: PieceType, case: PieceCycle, algorithms: Set<Algorithm>): Boolean {
        return algorithms.fold(true) { acc, alg -> acc && this.addAlgorithm(type, case, alg) }
    }

    fun mayUpdate(): Boolean

    fun updateAlgorithm(type: PieceType, oldAlg: Algorithm, newAlg: Algorithm): Boolean

    fun mayDelete(): Boolean

    fun deleteAlgorithm(type: PieceType, algorithm: Algorithm): Boolean

    fun deleteAlgorithms(type: PieceType, case: PieceCycle): Boolean
}
